### ISSUE TITLE
Expose StepPath parameter in Import-GherkinSteps

### ISSUE THREAD

**Issue by mirogta** [pester/Pester#1341]
## 1. General summary of the issue

I have a use case where I want different steps (`*.Steps.ps1`) to be executed for a single feature (`*.feature`). I have found out that there is a `StepPath` parameter in `Import-GherkinSteps` function, but there is no way to set the parameter when calling `Invoke-Gherkin`.

I've described the use case here: https://www.linkedin.com/pulse/how-did-i-implement-windows-server-hardening-cis-benchmark-sommer/

## 2. Describe Your Environment

Pester 4.8.1

## 3. Expected Behavior

I'd like to run `Invoke-Gherkin -StepsPath $StepsPath`

## 4.Current Behavior

There is currently no way to explicitly set the StepsPath, it's based on the path of feature files.

## 5. Possible Solution

Assuming this is potentially a rare use case, would it work with setting up a new environment variable like `PSPesterStepsPath`? If set, it could override the default StepsPath.

## 6. Context

How did this affect me? I've written a quick fix which modifies the Gherkin.ps1 code and injects a code which picks up a desired StepPath, runs Invoke-Gherking, and reverts the modification back to the original code. I'd like to remove this quick fix from my code to make it clean.

What am I trying to accomplish? Run test steps from a different directory for the same feature files, based on some conditions. See the link above.

**Comment by fourpastmidnight** [comment]
This is by design. The Gherkin runner will load all *.Steps.ps1 files _below_ the current directory (passed to the `-Path` parameter), as this is how Ruby Cucumber works, which is what the Gherkin runner is modeled after.

If you want different steps to run for a single feature, it sounds to me like you really have two different features. Perhaps the application feature that you're testing may be the same feature, but in each set of steps you're using a different strategy to execute the application feature? In this case, you're not testing the application feature so much as you are the strategy used to execute the feature. In this case, perhaps you can re-write the features and scenarios around the strategy being used? This would make it clear that you're really testing the strategies involved in orchestrating some feature of your application. (This doesn't mean you don't test the application feature in some way&mdash;just that testing of the application feature is orthogonal to the testing of the strategies used to orchestrate the feature.)

Oftentimes in Cucumber/Gherkin, you need to circle back around and ask, "What am I _really_ testing here?" And through this introspection, you may find that what you thought you were testing is not really what you intended to test at all. 😄

**Comment by mirogta** [comment]
Sure, I understand.

Have you read my article?

I'm not necessarily testing a strategy used to execute the feature... I have a feature file which describes some scenarios, one set of steps in one directory which implements them (changes the system settings), and another set of steps in another directory which validates them (validates the system settings). It's crucial that these steps are driven off the same feature file - it won't make any sense to have duplicate feature files.

We can debate whether Pester/Gherkin _should be only used_ to run tests and _should not be used_ to execute a code which changes the system. In my use case it made a perfect sense and helped to deliver a project and make our customers happy. Security experts who didn't know PowerShell could easily contribute to writing the feature files, because they're in plain English and so simple to write.

But I had to fork and hack one line in Pester's source code to make this possible and it would be much neater if it's supported without hacks. Hence raising this issue.

**Comment by fourpastmidnight** [comment]
I did read your issue&mdash;in full, twice, maybe even three times. But obviously, I'm missing _a lot_ of context since I can't see the structure of your tests and what you're trying to do. 😄 

I think part of the issue here is the overloaded use of the word "feature". So there's an application "feature", a use case if you will. Then there are Gherkin/Cucumber features, represented with feature files.

Having thought about your situation a bit more, it seems like you have something similar to this happening:

```
// setup some application state
// execute some stuff
// assert some stuff happened
//
// use state of application from previous "test"
// execute some more stuff
// assert some other stuff happened
```

_Usually_ this is not a good testing practice&mdash;but of course, everything in software is "it depends". Since I cannot see what exactly you're trying to accomplish, I'm going to trust that your situation warrants this.

I wonder if it's possible for you to split this up into two different tests? Perhaps you could use a `Scenario Outline` for this? Could you do something like this:

```
// setup the application state
// execute some stuff
// assert some stuff happened

// setup the application state as if the previous test was run
// execute some stuff
// assert other stuff happened
```

Then you've clearly broken this up into two tests. Since you indicated that you're using the same feature file for both tests, you might be able to use a `Scenario Outline`:

```gherkin
Feature: Adding two numbers
  As a user
  I want to be able to do simple math
  So I don't make dumb mistakes

  Scenario Outline: Can add two numbers
    Given two numbers <x> and <y>
    When I add them together
    Then I should get <result>

    Examples: Adding two numbers
      | x | y | result |
      | 1 | 1 | 2      |
      | 2 | 2 | 4      |
      | 2 | 3 | 5      |
```

In the example above, each step in the scenario above are run for each row in the table.

I'm not sure if this would help in your situation or not.

Please remember, I'm just trying to help. Usually there's a way to accomplish what you're trying to do with Cucumber/Gherkin just the way it is. Sometimes it just takes having a different perspective or approach.

**Comment by fourpastmidnight** [comment]
I remembered just now the scenario outline probably won't work in this case. Each scenario executes in its own world, not remembering the state from previous scenarios. And this is a good thing. Unit tests (and even integration tests) should be isolated from one another.

**Comment by mirogta** [comment]
@fourpastmidnight It's a good feedback - thank you
